processing/src/main/java/io/druid/query/filter/BoundDimFilter.java

```diff
@@ -312,114 +318,198 @@ public String toString()
-          private final long upperLongBound = hasUpperLongBound ? BoundLongPredicateSupplier.this.upperLongBound : 0L;
-
-          @Override
-          public boolean applyLong(long input)
-          {
-            int lowerComparing = 1;
-            int upperComparing = 1;
-            if (hasLowerLongBound) {
-              lowerComparing = Long.compare(input, lowerLongBound);
-            }
-            if (hasUpperLongBound) {
-              upperComparing = Long.compare(upperLongBound, input);
-            }
-            if (lowerStrict && upperStrict) {
-              return ((lowerComparing > 0)) && (upperComparing > 0);
-            } else if (lowerStrict) {
-              return (lowerComparing > 0) && (upperComparing >= 0);
-            } else if (upperStrict) {
-              return (lowerComparing >= 0) && (upperComparing > 0);
-            }
-            return (lowerComparing >= 0) && (upperComparing >= 0);
-
-          }
-        };
+        initPredicate();
+        return predicate;
       }
 
-      private void initLongData()
+      private void initPredicate()
       {
-        if (longsInitialized) {
+        if (predicate != null) {
           return;
         }
 
         synchronized (initLock) {
-          if (longsInitialized) {
+          if (predicate != null) {
             return;
           }
 
+          final boolean hasLowerLongBound;
+          final boolean hasUpperLongBound;
+          final long lowerLongBound;
+          final long upperLongBound;
           matchesNothing = false;
 
           if (hasLowerBound()) {
             final Long lowerLong = GuavaUtils.tryParseLong(lower);
             if (lowerLong == null) {
               // Unparseable values fall before all actual numbers, so all numbers will match the lower bound.
               hasLowerLongBound = false;
+              lowerLongBound = 0L;
             } else {
               hasLowerLongBound = true;
               lowerLongBound = lowerLong;
             }
           } else {
             hasLowerLongBound = false;
+            lowerLongBound = 0L;
           }
 
           if (hasUpperBound()) {
             Long upperLong = GuavaUtils.tryParseLong(upper);
             if (upperLong == null) {
               // Unparseable values fall before all actual numbers, so no numbers can match the upper bound.
               matchesNothing = true;
-              return;
+              hasUpperLongBound = false;
+              upperLongBound = 0L;
+            } else {
+              hasUpperLongBound = true;
+              upperLongBound = upperLong;
             }
-
-            hasUpperLongBound = true;
-            upperLongBound = upperLong;
           } else {
             hasUpperLongBound = false;
+            upperLongBound = 0L;
           }
 
-          longsInitialized = true;
+          if (matchesNothing) {
+            predicate = DruidLongPredicate.FALSE_PREDICATE;
+          } else {
```

Comment:
I've split this to create a separate predicate for each of the possible combinations